tools/federation/src/main/java/org/eclipse/rdf4j/federated/optimizer/LimitOptimizer.java

```diff
@@ -65,6 +65,18 @@ public void meet(Slice node) throws OptimizationException {
 			applicableLimitInScope = node.getLimit();
 		}
 		super.meet(node);
+
+		TupleExpr expr = node.getArg();
+		// if the top most element is a statement (e.g. for an ASK query with single statement pattern),
+		// i.e. no join, union or
+		// any other complex pattern, we can push the limit
+		// => this case typically represents a query with a single BGP
+		if (expr instanceof FedXStatementPattern) {
```

Comment:
this is applying the same logic as a few lines below.

The difference is that in case of a SELECT query, we find the single statement pattern below a `Projection` node and not directly as argument of the `Slice`